website/addons/mendeley/model.py

```diff
@@ -0,0 +1,335 @@
+
+    _api = None
+
+    @property
+    def api(self):
+        """authenticated ExternalProvider instance"""
+        if self._api is None:
+            self._api = Mendeley()
+            self._api.account = self.external_account
+        return self._api
+
+    @property
+    def has_auth(self):
+        return self.external_account is not None
+
+    @property
+    def selected_folder_name(self):
+        if self.mendeley_list_id is None:
+            return ''
+        elif self.mendeley_list_id != 'ROOT':
+            folder = self.api._folder_metadata(self.mendeley_list_id)
+            return folder.name
+        else:
+            return 'All Documents'
+
+    def grant_oauth_access(self, user, external_account, metadata=None):
+        """Grant OAuth access, updates metadata on user settings
+        :param User user:
+        :param ExternalAccount external_account:
+        :param dict metadata:
+        """
+        user_settings = user.get_addon('mendeley')
+
+        # associate the user settings with this node's settings
+        if user_settings not in self.associated_user_settings:
+            self.associated_user_settings.append(user_settings)
+            self.save()
+
+        user_settings.grant_oauth_access(
+            node=self.owner,
+            external_account=external_account,
+            metadata=metadata or {},
+        )
+
+        user_settings.save()
+
+    def verify_oauth_access(self, external_account, list_id):
+        """Determine if access to the ExternalAccount has been granted
+        :param ExternalAccount external_account:
+        :param str list_id: ID of the Mendeley list requested
+        :rtype bool:
+        """
+        for user_settings in self.associated_user_settings:
+            try:
+                granted = user_settings.oauth_grants[self.owner._id][
+                    external_account._id]
+            except KeyError:
+                # no grant for this node, move along
+                continue
+
+            if list_id in granted.get('lists', []):
+                return True
+        return False
+
+    def get_accounts(self, user):
+        accounts = [
+            account for account
+            in user.external_accounts
+            if account.provider == 'mendeley'
+        ]
+        if self.external_account and self.external_account not in accounts:
+            accounts.append(self.external_account)
+        return accounts
+
+    def to_json(self, user):
+        ret = super(AddonMendeleyNodeSettings, self).to_json(user)
+        ret.update({
+            'listId': self.mendeley_list_id,
+            'accounts': [utils.serialize_account(each) for each in
+                         self.get_accounts(user)],
+            'currentAccount': utils.serialize_account(self.external_account),
+        })
+        return ret
+
+
+class Mendeley(ExternalProvider):
+    name = "Mendeley"
+    short_name = "mendeley"
+
+    client_id = settings.MENDELEY_CLIENT_ID
```

Comment:
This should come from `website.addons.mendeley.settings`, not `website.settings`.